### PR TITLE
Swift TOO Form

### PR DESCRIPTION
**Pull request by mcoughlin**
A first pass at a Swift TOO Form. Not sure how best to test it... they do have an anonymous account:
https://www.swift.psu.edu/too_api/index.php?md=README.md
that let's you validate but not submit.

Closes https://github.com/skyportal/skyportal/issues/2422

### CODE REVIEW

**Review by profjsb — COMMENTED** [review]
I'm confused by the UVOT-centric naming (`UVOTAPI`) but most of the data payload seems to be XRT focused. Worth making different requests for different Swift instruments?

As for testing, since the end point will be brittle, we should probably use a mock return using `vcr`: https://github.com/ktosiek/pytest-vcr

File: skyportal/facility_apis/swift.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        too.exp_time_per_visit = request.payload["exposure_time"]
+        too.monitoring_freq = "%d days" % request.payload["monitoring_freq"]
+        too.num_of_visits = int(request.payload["exposure_counts"])
+
+        too.xrt_countrate = request.payload["xrt_countrate"]
+        too.exp_time_just = request.payload["exp_time_just"]
+        too.immediate_objective = request.payload["immediate_objective"]
+
+        too.urgency = int(request.payload["urgency"])
+        too.obs_type = request.payload["obs_type"]
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/facility_apis/swift.py:69]: It is a dropdown menu that restricts to:
            "obs_type": {
                "type": "string",
                "enum": ["Spectroscopy", "Light Curve", "Position", "Timing"],
                "default": "Light Curve",
                "title": "Observation Type",
            },
currently.

File: skyportal/facility_apis/swift.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+        too.exp_time_per_visit = request.payload["exposure_time"]
+        too.monitoring_freq = "%d days" % request.payload["monitoring_freq"]
+        too.num_of_visits = int(request.payload["exposure_counts"])
+
+        too.xrt_countrate = request.payload["xrt_countrate"]
+        too.exp_time_just = request.payload["exp_time_just"]
+        too.immediate_objective = request.payload["immediate_objective"]
+
+        too.urgency = int(request.payload["urgency"])
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/facility_apis/swift.py:68]: https://www.swift.psu.edu/too_api/index.php?md=TOO%20parameters.md
I think it is 1-4.

**Comment by mcoughlin** [comment]
My understanding is people just request both instruments when they put in the form, maybe we call it UVOT_XRT or something given it is just this one pathway? I could use help from someone on the VCR stuff if it is decided that is important.

**Comment by mcoughlin** [comment]
@profjsb the VCR wasn't too bad.

File: skyportal/facility_apis/swift.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (acrellin) [reply skyportal/facility_apis/swift.py:68]: Looks like we still need to validate the inputs here

[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/facility_apis/swift.py:68]: What would this look like? Don't we in essence validate at the step where we are giving input options in the dropdown?

[the remark this answers is not in the document]
Comment (profjsb) [reply skyportal/facility_apis/swift.py:68]: A key idea is that the frontend and the backend need to independently validate and verify inputs. Not doing so on the backend opens up the API to malformed inputs (ie. if someone decides to script again this endpoint). So even if the frontend is constraining the inputs we need to assume the backend has to do the same checks.

File: skyportal/facility_apis/swift.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (profjsb) [reply skyportal/facility_apis/swift.py:69]: See comment above--still need the backend to validate the inputs.

File: skyportal/facility_apis/swift.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (profjsb) [reply skyportal/facility_apis/swift.py:68]: See lines 88-134 in skyportal/handlers/api/internal/standards.py for range and type checking.

File: skyportal/facility_apis/swift.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/facility_apis/swift.py:79]: I added some validation as requested.

**Comment by mcoughlin** [comment]
@acrellin worth a look when you can.

File: skyportal/facility_apis/swift.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        """
+
+        self.requestgroup = self._build_payload(request)
+
+    def _build_payload(self, request):
+        """Payload json for Swift UVOT TOO requests.
+
+        Parameters
+        ----------
+
```
Comment (acrellin) [inline skyportal/facility_apis/swift.py:38]:
```suggestion
```

File: skyportal/facility_apis/swift.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+class UVOTRequest:
+
+    """A JSON structure for Swift UVOT requests."""
+
+    def __init__(self, request):
+        """Initialize UVOT request.
+
+        Parameters
+        ----------
+
```
Comment (acrellin) [inline skyportal/facility_apis/swift.py:25]:
```suggestion
```

File: skyportal/facility_apis/swift.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+        """Payload json for Swift UVOT TOO requests.
+
+        Parameters
+        ----------
+
+        request: skyportal.models.FollowupRequest
+            The request to add to the queue and the SkyPortal database.
+
+        Returns
+        ----------
```
Comment (acrellin) [inline skyportal/facility_apis/swift.py:43]:
```suggestion
        Returns
        -------
```

File: skyportal/facility_apis/swift.py [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+            raise ValueError('Missing allocation information.')
+
+        swiftreq = UVOTRequest(request)
+
+        swiftreq.requestgroup.validate()
+
+        r = requests.post(
+            url=API_URL, verify=True, data={'jwt': swiftreq.requestgroup.jwt}
+        )
+        # r = swiftreq.requestgroup.submit_post()
```
Comment (acrellin) [inline skyportal/facility_apis/swift.py:116]:
```suggestion
```

File: skyportal/facility_apis/swift.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+            request.status = f'rejected: {r.content}'
+
+        transaction = FacilityTransaction(
+            request=http.serialize_requests_request(r.request),
+            response=http.serialize_requests_response(r),
+            followup_request=request,
+            initiator_id=request.last_modified_by_id,
+        )
+
+        DBSession().add(transaction)
```
Comment (acrellin) [inline skyportal/facility_apis/swift.py:130]: Need to commit afterwards?

File: skyportal/tests/frontend/sources_and_followup_etc/test_followup_requests.py [hunk 8]
```diff
@@ -52,6 +52,16 @@
     ztf_isonline = True
 
 
+url = "https://www.swift.psu.edu/toop/submit_json.php"
+swift_isonline = False
+try:
+    requests.get(url, timeout=5)
```
Comment (acrellin) [inline skyportal/tests/frontend/sources_and_followup_etc/test_followup_requests.py:58]:
```suggestion
swift_url = "https://www.swift.psu.edu/toop/submit_json.php"
swift_isonline = False
try:
    requests.get(swift_url, timeout=5)
```

**Review by acrellin — COMMENTED** [review]
Looks good, thanks! A few questions below

File: data/tests/test_server_recordings_static.yaml [hunk 9]
```diff
@@ -240,4 +240,48 @@ interactions:
...
...0IiwiYXBpX3ZlcnNpb24iOiIxLjIiLCJhcGlfZGF0YSI6eyJ1c2VybmFtZSI6ImFub255bW91cyIsInNvdXJjZV9uYW1lIjoiMTc1ZTZkMjNiMWEzNDM4YTljYjA1YTg2YzBmYzM2NTQiLCJzb3VyY2VfdHlwZSI6Ik9wdGljYWwgZmFzdCB0cmFuc2llbnQiLCJyYSI6MC4wLCJkZWMiOjAuMCwicG9zZXJyIjowLjAsImluc3RydW1lbnQiOiJYUlQiLCJ1cmdlbmN5IjozLCJ4cnRfY291bnRyYXRlIjowLjAwMjUsImltbWVkaWF0ZV9vYmplY3RpdmUiOiJXZSB3aXNoIHRvIG1lYXN1cmUgdGhlIFgtcmF5IGVtaXNzaW9uIG9mIGFuIG9wdGljYWxseSBkaXNjb3ZlcmVkIHBvdGVudGlhbCBvcnBoYW4gYWZ0ZXJnbG93L2tpbG9ub3ZhLiIsInNjaWVuY2VfanVzdCI6IkFuIFgtcmF5IGRldGVjdGlvbiBvZiB0aGlzIHRyYW5zaWVudCB3aWxsIGZ1cnRoZXIgYXNzb2NpYXRlIHRoaXMgb2JqZWN0IHRvIGEgcmVsYXRpdmlzdGljIGV4cGxvc2lvbiBhbmQgd2lsbCBoZWxwIHVudmVpbCB0aGUgbmF0dXJlIG9mIHRoZSBwcm9nZW5pdG9yIHR5cGUuIiwiZXhwb3N1cmUiOjQwMDAsImV4cF90aW1lX2p1c3QiOiJBdCB-Mi41ZS0zIGNvdW50cy9zZWMsIDRrcyBzaG91bGQgc3VmZmljZSB0byBhY2hpZXZlIGEgaGlnaCBTTlIsIGFzc3VtaW5nIGEgYmFja2dyb3VuZCBvZiB-MWUtNCBjb3VudHMvc2VjIChQYWdhbmkgZXQgYWwuIDIwMDcpIiwiZXhwX3RpbWVfcGVyX3Zpc2l0Ijo0MDAwLCJudW1fb2ZfdmlzaXRzIjoxLCJtb25pdG9yaW5nX2ZyZXEiOiIxIGRheXMiLCJ4cnRfbW9kZSI6IlBDIiwidXZvdF9tb2RlIjoiZGVmYXVsdCIsIm9ic19uIjoic2luZ2xlIiwib2JzX3R5cGUiOiJMaWdodCBDdXJ2ZSIsInZhbGlkYXRlX29ubHkiOmZhbHNlfX0.PrzYHA42kdg5apdpJD_7xRn7RsW0eGHGj3TDTw0jN5c
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply data/tests/test_server_recordings_static.yaml:244]: It is just the anonymous account they suggest for testing.

File: skyportal/facility_apis/swift.py (hunk 7 above)
Comment (mcoughlin) [reply skyportal/facility_apis/swift.py:127]: This is the same as in the others... it clearly appears on the page, so I presume it works. But I cannot tell you why. This was a Danny invention...

Comment (acrellin) [reply skyportal/facility_apis/swift.py:127]: Okay, we must be calling `commit` after this method is invoked.

**Review by acrellin — APPROVED** [review]

**Review by profjsb — APPROVED** [review]
Looks great. Will merge after tests pass.

**Comment by profjsb** [comment]
@mcoughlin would you mind merging current master into this branch (I dont have permission)? I'll rerun tests then.

**Comment by mcoughlin** [comment]
Rebased

File: skyportal/facility_apis/swift.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+        """Payload json for Swift UVOT TOO requests.
+
+        Parameters
+        ----------
+        request: skyportal.models.FollowupRequest
+            The request to add to the queue and the SkyPortal database.
+
+        Returns
+        -------
+        payload: json
```
[the remark this answers is not in the document]
Comment (stefanv) [reply skyportal/facility_apis/swift.py:42]: The return type here doesn't seem to be JSON?

File: skyportal/facility_apis/swift.py [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+            payload for requests.
+        """
+
+        altdata = request.allocation.altdata
+
+        too = Swift_TOO()
+        too.username = altdata["username"]
+        too.shared_secret = altdata["secret"]
+        # too.debug = True
```
Comment (stefanv) [inline skyportal/facility_apis/swift.py:51]: Remove?

File: skyportal/facility_apis/swift.py [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+
+    # subclasses *must* implement the method below
+    @staticmethod
+    def submit(request):
+
+        """Submit a follow-up request to Swift's UVOT
+
+        Parameters
+        ----------
+        request: skyportal.models.FollowupRequest
```
Comment (stefanv) [inline skyportal/facility_apis/swift.py:97]:
```suggestion
        request : skyportal.models.FollowupRequest
```

File: skyportal/facility_apis/swift.py [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+
+        altdata = request.allocation.altdata
+        if not altdata:
+            raise ValueError('Missing allocation information.')
+
+        swiftreq = UVOTRequest(request)
+
+        swiftreq.requestgroup.validate()
+
+        r = requests.post(
```
Comment (stefanv) [inline skyportal/facility_apis/swift.py:111]: Handle exceptions gracefully. https://docs.python-requests.org/en/latest/user/quickstart/#errors-and-exceptions

File: skyportal/services/test_server/test_api_server.py [hunk 14]
```diff
@@ -390,9 +414,12 @@ def get(self):
                 self.write("Could not find test route redirect")
 
     def post(self):
+
         is_soap_action = "Soapaction" in self.request.headers
         if "/api/requestgroups/" in self.request.uri:
             cache = get_cache_file_static()
+        elif "/toop/submit_json.php" in self.request.uri:
```
Comment (stefanv) [inline skyportal/services/test_server/test_api_server.py:421]: Perhaps something like:

```
cached_urls = [".*/api/requestgroups/.*", ".*/toop/submit_json.php$"]

...

if any(re.match(pat, self.request.uri) for pat in cached_urls):
    cache = get_cache_file_static()
```

File: skyportal/tests/frontend/sources_and_followup_etc/test_followup_requests.py [hunk 15]
```diff
@@ -52,6 +52,16 @@
     ztf_isonline = True
 
 
+swift_url = "https://www.swift.psu.edu/toop/submit_json.php"
```
Comment (stefanv) [inline skyportal/tests/frontend/sources_and_followup_etc/test_followup_requests.py:55]: Do we need to test against swift if we have the pre-recorded transactions?

File: skyportal/utils/http.py [hunk 16]
```diff
@@ -1,6 +1,9 @@
 def serialize_requests_request(request):
     if request.body is not None:
-        body = request.body.decode()
+        if isinstance(request.body, bytes):
```
Comment (stefanv) [inline skyportal/utils/http.py:3]: What's happening here?

**Review by stefanv — COMMENTED** [review]
Nice work, @mcoughlin!

Some minor suggestions, hopefully all very quick.

File: skyportal/utils/http.py (hunk 16 above)
Comment (mcoughlin) [reply skyportal/utils/http.py:3]: just checking if needs to be decoded or not.

File: skyportal/tests/frontend/sources_and_followup_etc/test_followup_requests.py (hunk 15 above)
Comment (mcoughlin) [reply skyportal/tests/frontend/sources_and_followup_etc/test_followup_requests.py:55]: We do the others...

File: skyportal/services/test_server/test_api_server.py (hunk 14 above)
Comment (mcoughlin) [reply skyportal/services/test_server/test_api_server.py]: added

File: skyportal/facility_apis/swift.py (hunk 13 above)
Comment (mcoughlin) [reply skyportal/facility_apis/swift.py:110]: added

File: skyportal/facility_apis/swift.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/facility_apis/swift.py]: fixed

File: config.yaml.defaults [hunk 17]
```diff
@@ -33,6 +33,10 @@ app:
     host: skipper.caltech.edu
     port: 4000
 
+  swift_protocol: https
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply config.yaml.defaults]: fixed

File: data/tests/test_server_recordings_static.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (mcoughlin) [reply data/tests/test_server_recordings_static.yaml:244]: Kyung just wrote a note at the top of the script to delete the file and rerun the tests.

File: skyportal/utils/http.py (hunk 16 above)
Comment (stefanv) [reply skyportal/utils/http.py:3]: Yes, but why would it ever arrive as anything other than bytes?

Comment (mcoughlin) [reply skyportal/utils/http.py]: Ok will remove it

Comment (stefanv) [reply skyportal/utils/http.py]: I presume it was added for a reason; just trying to understand why!

Comment (mcoughlin) [reply skyportal/utils/http.py]: Presumably some bug I introduced in initial testing.

Comment (mcoughlin) [reply skyportal/utils/http.py:3]: Well that broke the tests, so something about it is important I guess

Comment (stefanv) [reply skyportal/utils/http.py:3]: Hrm, it is used by sedm, lt, ztf, and lco without this switch.  Why would swift be different?

Comment (stefanv) [reply skyportal/utils/http.py:3]: Looks like the failures are unrelated?

Comment (mcoughlin) [reply skyportal/utils/http.py:3]: I am not sure why. something about how the Swift folks are returning requests?

**Comment by mcoughlin** [comment]
@stefanv can we get this merged?

**Comment by profjsb** [comment]
@mcoughlin please rebase off of current master so we can get CI tests passing.

**Comment by mcoughlin** [comment]
@profjsb I think it is all rebased, according to github. Do you see something else?

**Comment by profjsb** [comment]
Since we're updating `facilities_apis` ENUM with UVOTAPI, you will need a migration script (see https://github.com/profjsb/skyportal/blob/90eccb64de9c68621f2ed25f61c9a3e22abca7d9/alembic/versions/f7c1e785d7a4_unique_token_name_by_user.py). Once you pull in the latest on master and run the tests you'll get migration.py and migra.sql artifact from the (initially failing) migration script. You'll combine those two files as above. Once you include that script you should get tests to pass

**Comment by mcoughlin** [comment]
@profjsb Doing this by hand I get an empty migration script (and github doesn't give me a migration).

File: requirements.txt [hunk 18]
```diff
@@ -51,3 +51,4 @@ pygcn==1.1.2
 xmlschema==1.9.0
 requests==2.27
 typing-extensions>=4.0.1  # astropy v5.0 needs this, but does not correctly specify as much in its install config
+swifttools==2.1
```
Comment (profjsb) [inline requirements.txt:54]:
```suggestion
swifttools==2.2.1
```

File: skyportal/facility_apis/swift.py [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+        too.exp_time_per_visit = request.payload["exposure_time"]
+        too.monitoring_freq = "%d days" % request.payload["monitoring_freq"]
+        too.num_of_visits = int(request.payload["exposure_counts"])
+
+        too.xrt_countrate = request.payload["xrt_countrate"]
+        too.exp_time_just = request.payload["exp_time_just"]
+        too.immediate_objective = request.payload["immediate_objective"]
+
+        if request.payload["urgency"] not in ["1", "2", "3", "4"]:
+            raise ValueError('urgency outside of allowed bounds.')
```
Comment (profjsb) [inline skyportal/facility_apis/swift.py:66]:
```suggestion
            raise ValueError('urgency not in ["1", "2", "3", "4"].')
```

File: skyportal/facility_apis/swift.py [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+        swiftreq = UVOTRequest(request)
+
+        swiftreq.requestgroup.validate()
+
+        r = requests.post(
+            url=API_URL, verify=True, data={'jwt': swiftreq.requestgroup.jwt}
+        )
+        r.raise_for_status()
+
+        if r.status_code == 201:
```
Comment (profjsb) [inline skyportal/facility_apis/swift.py:115]: not 200?

File: skyportal/facility_apis/__init__.py [hunk 21]
```diff
@@ -4,4 +4,5 @@
 from .sedmv2 import SEDMV2API
 from .lt import IOOAPI, IOIAPI, SPRATAPI
 from .lco import SINISTROAPI, SPECTRALAPI, FLOYDSAPI, MUSCATAPI
+from .swift import UVOTAPI
```
Comment (profjsb) [inline skyportal/facility_apis/__init__.py:7]: Suggest changing this to SWIFTAPI or UVOT_XRTAPI to get the naming consistent with the instruments that will be requested.

**Review by profjsb — CHANGES_REQUESTED** [review]
See https://github.com/skyportal/skyportal/commit/dba29ba7ecf836f8c1ce0aa2157764c450444ab0# to see the upgrade/downgrade path. You should see the commands in the migral.sql script which is an artifact of the migration GHA.

**Comment by profjsb** [comment]
@stefanv can you finalize your review here? I think that the reason that the migration GHA is not getting triggered is because we dont check the change of `skyportal/facility_apis/__init__.py`. That said, I think that the refresh enums code will take care of the changing enums, right? If not, then we need to trigger a migration on enum change.

File: skyportal/facility_apis/swift.py (hunk 20 above)
Comment (mcoughlin) [reply skyportal/facility_apis/swift.py:115]: this is  what they send back...

File: skyportal/facility_apis/__init__.py (hunk 21 above)
Comment (mcoughlin) [reply skyportal/facility_apis/__init__.py]: I changed it to UVOTXRTAPI.

**Comment by profjsb** [comment]
@mcoughlin after you fix the migration to the latest revision we can merge.

**Comment by mcoughlin** [comment]
@profjsb it should be latest now. I updated this morning.

**Review by profjsb — APPROVED** [review]

**Pull request merged**